enhancements/cluster-runtime-constraints.md

```diff
@@ -0,0 +1,138 @@
+
+## Motivation
+
+At the moment, OLM resolves a set of operators which seem that they will work together based on dependency requirements that are specified by operator bundles. However, it is possible that those resolved operators will not work on the clusters due to cluster/runtime constraints. For example, if the cluster is running a specific Kubernetes version that is not compatible with the operators, the installed operators may fail to work properly on that given cluster.
+
+The cluster runtime constraints are often unknown by the operator authors and should be specified by cluster admins. However, currently, there is no mechanism for cluster admins to specify runtime constraints that OLM can understand and take under consideration during installation.
+
+### Goals
+- Provide mechanism (including specification and guideline) for cluster admins to specify cluster runtime constraints
+- Provide specification and guideline for author operators to specify runtime constraints/requirements if needed
+- Block the updates/installations of operators based on cluster runtime constraints
+- Allow the cluster runtime constraints to be generic and configurable by cluster admins
+
+### Non-Goals
+
+- Provide mechanism to retrieve/populate cluster runtime constraints automatically
+- Allow cluster runtime constraints to be scoped
+- Allow cluster runtime constraints to come from multiple/different sources
+- Allow optional cluster runtime constraints
+
+
+## Proposal
+
+To provide a generic solution for cluster runtime constraints, a library named [cel-go](https://github.com/google/cel-go) that provides expression evaluation capabilities is utilized in this proposal.
+
+### User Stories
+
+#### Constraints specified by cluster admin
+
+As a cluster admin, I would like to specify a list of cluster runtime constraints that are associated with a given cluster and all installable operators will meet these requirements.
+
+#### Constraint properties specified by operator author
+
+As an operator author, I would like to provide a list of runtime properties that may be needed to be considered before installation to ensure the operator working properly on the cluster.
+
+### Implementation Details
+
+#### Cluster Runtime Constraint Configmap
+
+Cluster admins can specified cluster runtime constraints in a configmap (named `olm-runtime-constraints`) in `olm` namespace using this format:
+
+```yaml=
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: olm-runtime-constraints
+  namespace: olm
+data:
+    properties: <list-of-expression-constraints-in-json>
+```
+
+Each constraint is specified using the following syntax:
+
+```json=
+{
+	"type": "olm.constraint",
+	"value": {
+		"evaluator": {
+			"id": "cel"
+		},
+		"source": "properties.exists(p, p.type == \"certified\")",
```

Comment:
Am I understanding correctly that this means operator authors and cluster admins have agreed out-of-band that "type" is a property with certain meaning, and that "certified" is a possible value with a certain meaning?

Using the CPU count example cited earlier, how can operator authors and cluster admins coordinate a name for that property? If 10 different operator projects all care about CPU count, how do you prevent one project from calling it "cpus", another calling it "cpu_count", another "vcpus", another "cores", etc.? Is there some central authority on what the property names should be, their data type, possible values, etc?

Reply:
Even with a CRD, if I'm understanding what you're implying, don't you still have the problem that when operator authors write down their operators' requirements in some bundle metadata, they have to choose field names, and different authors might use different names?

Reply:
There's another downside to the openendedness of this design.

If there's a typo in the property `type`, it seems like this design would just ignore that property without any obvious indication to the user.

For example, say my cluster has a valid constraint against a property called `olm.supportedOpenShiftVersionRange`, and I have a bundle that includes that property, but accidentally mistypes it as `olm.supportedOpenshiftVersionRange` (lowercase 's' in Openshift, instead of uppercase)

The constraint on the cluster would not apply to this bundle because it doesn't actually have the property we're looking for, and the bundle would be an installation/upgrade candidate even when the intent was the opposite. 

The same problem is true in reverse for constraint types. If I accidentally type `olm.contraint` (leave out an 's'), all of a sudden, that looks like an opaque property.

As far as I can tell, there would be no easy way for OLM to surface to the user "hey, you're `olm.supportedOpenshiftVersionRange` looks off". It's hard enough _right now_ for users to debug resolution decisions. Just imagine how much harder it will be when clusters and bundles can start introducing orders of magnitude more inputs into resolution decisions.